### PR TITLE
Fix Gather operator when indices are negative

### PR DESCRIPTION
**Pull request by mateusztabaka**

### CODE REVIEW

**Comment by GlebKazantaev** [comment]
@dmitry-gorokhov  please look at this changes.

File: inference-engine/src/mkldnn_plugin/nodes/gather.cpp [hunk 1]
```diff
@@ -119,7 +119,9 @@ class GatherImpl: public ExtLayerBase {
         size_t len = dataLength * dictionary->getTensorDesc().getPrecision().size();
 
         parallel_for(src_indexSize, [&](size_t i) {
-            unsigned int idx = Conversion()(src_index[i]);
+            int idx = Conversion()(src_index[i]);
+            if (idx < 0)
+                idx += indexRange;
```
Comment (dmitry-gorokhov) [inline inference-engine/src/mkldnn_plugin/nodes/gather.cpp:124]: @mateusztabaka  Maybe I missed smt but the [spec ](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/movement/Gather_1.md) says that: 

> 2: Tensor with indices to gather. The values for indices are in the range [0, input1[axis] - 1]. Required.

From my understanding this means that negative index is invalid case from operation specification standpoint. Why do we need to change plug-in implementation in that case?

@lazarevevgeny Could you please help us to make it clear?

Comment (mateusztabaka) [reply inference-engine/src/mkldnn_plugin/nodes/gather.cpp:124]: > @mateusztabaka Maybe I missed smt but the [spec ](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/movement/Gather_1.md) says that:
> 
> > 2: Tensor with indices to gather. The values for indices are in the range [0, input1[axis] - 1]. Required.
> 
> From my understanding this means that negative index is invalid case from operation specification standpoint. Why do we need to change plug-in implementation in that case?

Sorry, I missed that in OV spec.
Negative indices is a requirement coming from onnx spec: https://github.com/onnx/onnx/blob/master/docs/Operators.md#Gather so that was my motivation of changing plugin implementation.

Comment (lazarevevgeny) [reply inference-engine/src/mkldnn_plugin/nodes/gather.cpp:124]: OpenVINO does not support negative indices on purpose. 

@mateusztabaka, have you seen any real ONNX model which contains Gather with negative indices?

Comment (mateusztabaka) [reply inference-engine/src/mkldnn_plugin/nodes/gather.cpp:124]: > OpenVINO does not support negative indices on purpose.
May I know the reason?

> @mateusztabaka, have you seen any real ONNX model which contains Gather with negative indices?
Not so far. However, one of the cases in onnx's testsuite (https://github.com/onnx/onnx/blob/master/onnx/backend/test/case/node/gather.py#L46) depends on that feature.

Comment (lazarevevgeny) [reply inference-engine/src/mkldnn_plugin/nodes/gather.cpp:124]: When this operation was added to the OpenVINO it was implemented in TensorFlow [only](https://www.tensorflow.org/api_docs/python/tf/gather). In TF specification negative indices are not supported. Also, due to performance reasons we don't allow negative indices for ScatterXXX operations.

Comment (mateusztabaka) [reply inference-engine/src/mkldnn_plugin/nodes/gather.cpp:124]: The reason to introduce negative indexes in onnx is explained here: https://github.com/onnx/onnx/pull/2260.
Quote:
> We would need to support negative indices for these operators, since PyTorch does support this case. Using Select/Gather ops with -1 index is quite common among PyTorch models.
